### ISSUE TITLE
Path towards removing Tensorflow

### ISSUE THREAD

**Issue by makortel** [cms-sw/cmssw#49678]
Tensorflow has been notoriously difficult external to deal with (see [issue search](https://github.com/search?q=repo%3Acms-sw%2Fcmssw+tensorflow&type=issues)). Latest episode was failures in some GPU unit tests with updated Eigen (that is coupled to Tensorflow version) in Python 3.12 (that also couples to Tensorflow version) builds https://github.com/cms-sw/cmssw/issues/49546.

We are also building Tensorflow with C++17, because TF 2.12 does not support any newer standard, and that risks ODR violations. The newer TF  2.17 in the PY312 IBs apparently would support C++20, but we have also started the (long) process to move to C++23, so it seems we can expect to continue with mismatch in supported C++ versions.

These last points raised (again) the question if all present uses of Tensorflow could be migrated to some other ML inference engine, and Tensorflow itself be removed. (I vaguely recall the removal of Tensorflow was discussed and "agreed" in some O&C week session, maybe in spring 2024, but I can't find anything from my notes)

**Comment by makortel** [comment]
assign core,ml

**Comment by makortel** [comment]
Here is a list of packages that depend on `PhysicsTools/TensorFlow` (ignoring itself and `PhysicsTools/TensorFlowAOT`)
```
DQM/DTMonitorClient
L1Trigger/L1CaloTrigger
L1Trigger/L1THGCal
L1Trigger/L1TMuonEndCap
L1Trigger/L1TMuonEndCapPhase2
L1Trigger/L1TTrackMatch
L1Trigger/Phase2L1ParticleFlow
L1Trigger/VertexFinder
PhysicsTools/BPHNano
PhysicsTools/PatAlgos
RecoEcal/EgammaCoreTools
RecoEgamma/EgammaPhotonProducers
RecoEgamma/EgammaTools
RecoEgamma/ElectronIdentification
RecoEgamma/PhotonIdentification
RecoHGCal/TICL
RecoMET/METPUSubtraction
RecoMuon/TrackerSeedGenerator
RecoTauTag/HLTProducers
RecoTauTag/RecoTau
RecoTracker/DisplacedRegionalTracking
RecoTracker/FinalTrackSelectors
RecoTracker/MkFit
RecoTracker/TkSeedGenerator
```

**Comment by makortel** [comment]
assign DQM/DTMonitorClient,L1Trigger/L1CaloTrigger,L1Trigger/L1THGCal,L1Trigger/L1TMuonEndCap,L1Trigger/L1TMuonEndCapPhase2,L1Trigger/L1TTrackMatch,L1Trigger/Phase2L1ParticleFlow,L1Trigger/VertexFinder,PhysicsTools/BPHNano,PhysicsTools/PatAlgos,RecoEcal/EgammaCoreTools,RecoEgamma/EgammaPhotonProducers,RecoEgamma/EgammaTools,RecoEgamma/ElectronIdentification,RecoEgamma/PhotonIdentification,RecoHGCal/TICL,RecoMET/METPUSubtraction,RecoMuon/TrackerSeedGenerator,RecoTauTag/HLTProducers,RecoTauTag/RecoTau,RecoTracker/DisplacedRegionalTracking,RecoTracker/FinalTrackSelectors,RecoTracker/MkFit,RecoTracker/TkSeedGenerator

**Comment by slava77** [comment]
is PyTorch supported?

**Comment by valsdav** [comment]
Hi Matti.  We confirmed that we agreed on the plan of trying to remove the TensorFlow dependecy by moving to ONNX of TF-AOT the existing models.  Unfortunately I couldn't find enough time or personpower to tackle that quickly, I'm sorry for that.   We propose to come back to the topic with a pragmatic plan to try to solve the issue as soon as possible.

We can split the TensorFlow usages in 3 cases:
1) simple inference calls with models that can be switched to ONNX (with the tf2onnx tool)  in a matter of few edits.
2) producers also using ESProducers to configure the models 
3) more convoluted producers using chains of TF models (e.g. DeepTau)

We propose to tackle directly (as ML production group) the cases 1 and 2, opening small PRs and preparing a simple ESRecord interface for ONNX as the one available for TensorFlow.  For case 3, we would need a bit more time or the help of the maintainers of the packages.

**Comment by makortel** [comment]
Thanks @valsdav!

**Comment by makortel** [comment]
> is PyTorch supported?

So some degree yes (I'd like @cms-sw/ml-l2 comment more on their side). IIRC PyTorch inference is not thread safe at the moment, which would imply the need to replicate the models to all framework streams, which then sounds potentially expensive in terms of memory (that is already a challenge).

**Comment by fwyzard** [comment]
tagging @waredjeb @felicepantaleo for `RecoHGCal/TICL`

**Comment by mmusich** [comment]
tagging @brallmond @kandrosov @valeriadamante for `RecoTauTag/HLTProducers/src/L2TauTagNNProducerAlpaka.cc`

EDIT:
there is a `TensorFlow` dependency *also* in `RecoTauTag/HLTProducers/src/L2TauTagNNProducer.cc`.
It's not clear to me if it still useful or given that we migrated everything to alpaka we can deprecate it? (I don't see it used anywhere at HLT).

**Comment by jfernan2** [comment]
Tagging:
@AuroraPerego @see-saw28 @cochando for RecoEGamma
@fsimone91 @llunerti  for RecoMuon
@lnurfikri89   for RecoMET
@ReyerBand for RecoEcal

**Comment by mmusich** [comment]
> EDIT:
> there is a TensorFlow dependency also in RecoTauTag/HLTProducers/src/L2TauTagNNProducer.cc.
> It's not clear to me if it still useful or given that we migrated everything to alpaka we can deprecate it? (I don't see it used anywhere at HLT).

the module got removed in https://github.com/cms-sw/cmssw/pull/49794.

**Comment by makortel** [comment]
To remind, here is the current list of packages depending on `PhysicsTools/TernsorFlow`
```
DQM/DTMonitorClient
L1Trigger/L1THGCal
L1Trigger/L1TMuonEndCap
L1Trigger/L1TMuonEndCapPhase2
L1Trigger/Phase2L1ParticleFlow
L1Trigger/VertexFinder
PhysicsTools/PatAlgos
RecoEcal/EgammaCoreTools
RecoHGCal/TICL
RecoMET/METPUSubtraction
RecoMuon/TrackerSeedGenerator
RecoTauTag/HLTProducers
RecoTauTag/RecoTau
RecoTracker/DisplacedRegionalTracking
RecoTracker/FinalTrackSelectors
RecoTracker/MkFit
RecoTracker/TkSeedGenerator
```

**Comment by jfernan2** [comment]
> DQM/DTMonitorClient removed in [#50136](https://github.com/cms-sw/cmssw/pull/50136)

@makortel I am confused, DQM/DTMonitorClient was removed in https://github.com/cms-sw/cmssw/pull/50136

Am I missing anything? Thanks

**Comment by makortel** [comment]
> > DQM/DTMonitorClient removed in [#50136](https://github.com/cms-sw/cmssw/pull/50136)
> 
> [@makortel](https://github.com/makortel) I am confused, DQM/DTMonitorClient was removed in [#50136](https://github.com/cms-sw/cmssw/pull/50136)
> 
> Am I missing anything? Thanks

Looks like the only thing left is
https://github.com/cms-sw/cmssw/blob/03018b9116f636ea1ce59b3af88bcacc161e0cc3/DQM/DTMonitorClient/BuildFile.xml#L14
(which is what I searched for with `git grep`)

**Comment by jfernan2** [comment]
Removed in https://github.com/cms-sw/cmssw/pull/50793

> > > DQM/DTMonitorClient removed in [#50136](https://github.com/cms-sw/cmssw/pull/50136)
> > 
> > 
> > [@makortel](https://github.com/makortel) I am confused, DQM/DTMonitorClient was removed in [#50136](https://github.com/cms-sw/cmssw/pull/50136)
> > Am I missing anything? Thanks
> 
> Looks like the only thing left is
> 
> [cmssw/DQM/DTMonitorClient/BuildFile.xml](https://github.com/cms-sw/cmssw/blob/03018b9116f636ea1ce59b3af88bcacc161e0cc3/DQM/DTMonitorClient/BuildFile.xml#L14)
> 
> Line 14 in [03018b9](/cms-sw/cmssw/commit/03018b9116f636ea1ce59b3af88bcacc161e0cc3)
>  <use name="PhysicsTools/TensorFlow"/> 
> 
> (which is what I searched for with `git grep`)

**Comment by mmasciov** [comment]
Just to make sure: is this targeting also Run-3 reconstruction? The question obviosuly stems from the fact that there are a few instances for tracking (@cms-sw/tracking-pog-l2) that still make use of Tensorflow in Run-3.

**Comment by makortel** [comment]
> is this targeting also Run-3 reconstruction?

I'd think we can leave Tensorflow to stay in CMSSW_17_0_X (i.e. Run 3 legacy). I hope we don't need to do (m)any major external updates there that would couple with Tensorflow.

Concerning the Run 4 targeting software (CMSSW_20_0_X and beyond), to be practical it's probably enough for the Tensorflow-using modules to be cleaned up as a part of the overall cleanup of Run 1-3 reconstruction, assuming that would happen during this year (the exact details on how long the Run 3 reconstruction needs to kept functional in the 20_Y are somewhat foggy).

**Comment by makortel** [comment]
Just to note here the latest hurdles connected to difficulties in updating Tensorflow https://github.com/cms-sw/cmsdist/issues/10716#issuecomment-5089111537

**Comment by IzaakWN** [comment]
Hi all,

I believe `flatbuffers` is currently a dependency of just Tensorflow. It is listed as an external in `cmsdist`: https://github.com/search?q=repo%3Acms-sw%2Fcmsdist%20flatbuffers&type=code. Could I kindly request `flatbuffers` is not removed from CMSSW externals?

HGCal BE is planning to use `flatbuffers` to encoding metadata (see [GitLab Work Item hgcal-dpg/hgcal-comm#87](https://gitlab.cern.ch/hgcal-dpg/hgcal-comm/-/work_items/87), [André's slides](https://indico.cern.ch/event/1634669/#49-introduction-of-metadata-in), [my slides](https://indico.cern.ch/event/1655746/timetable/?view=standard#72-hgcal)). This metadata will be stored in RAW data and has to be decoded again in CMSSW to steer configuration of scans and masking of channels. We have already tested it during the last test beam ([`HGCalMetaDataProducer.cc`](https://gitlab.cern.ch/hgcal-dpg/hgcal-comm/-/blob/master/SystemTestEventFilters/plugins/HGCalMetaDataProducer.cc?ref_type=heads#L7)). We would require anything newer than v2.0.8, which already seems to be the case since CMSSW 17_0.

Cheers,
Izaak

**Comment by makortel** [comment]
> This metadata will be stored in RAW data and has to be decoded again in CMSSW to steer configuration of scans and masking of channels.

Is this metadata something that would be stored in the `flatbuffers` format in the offline RAW that is expected to have infinite backwards compatibility guarantee?

**Comment by IzaakWN** [comment]
> > This metadata will be stored in RAW data and has to be decoded again in CMSSW to steer configuration of scans and masking of channels.
> 
> Is this metadata something that would be stored in the `flatbuffers` format in the offline RAW that is expected to have infinite backwards compatibility guarantee?

Hi @makortel, yes. This was a recent proposal to allow for efficient calibrations and monitoring without stopping a run. The current idea is to store the `flatbuffers` schema version in CondDB with an appropriate IoV. This would be used as a single source of truth for both the BE (to encode the metadata into RAW) and CMSSW (to decode the metadata from RAW). However, we expect this scheme to be stable during Phase II, just like the RAW (un)packing itself. 

HGCal BE and the HGCal Raw-data handling DPG plan to present this to the BE forum and CMSSW Core once it has been refined further. We are of course open to other suggestions if there are better solutions for seamless O2O use and long-term maintenance.

**Comment by mmusich** [comment]
>```
> RecoTauTag/RecoTau
> ```

I noticed by chance this DP note request [here](https://gitlab.cern.ch/cms-ppd/cms-ppd-matters/-/work_items/104).
May I ask TAU developers (@cms-sw/tau-pog-l2 @pritampalit) what is the plan here? 
- Will there be a cmssw PR to migrate `DeepTau` from TF to ONNX ?
- Is the general plan to still be supportive of `DeepTau` in the future (e.g. for Phase-2 HLT usage) in view of [this tentative DP note](https://indico.cern.ch/event/1722071/#6-part-ditau-path-in-phase-2-h)?

**Comment by mbluj** [comment]
> > ```
> > RecoTauTag/RecoTau
> > ```
> 
> I noticed by chance this DP note request [here](https://gitlab.cern.ch/cms-ppd/cms-ppd-matters/-/work_items/104). May I ask TAU developers ([@cms-sw/tau-pog-l2](https://github.com/orgs/cms-sw/teams/tau-pog-l2) [@pritampalit](https://github.com/pritampalit)) what is the plan here?
> 
> * Will there be a cmssw PR to migrate `DeepTau` from TF to ONNX ?

Yes, a PR is being worked on and should appear in a few weeks.

> * Is the general plan to still be supportive of `DeepTau` in the future (e.g. for Phase-2 HLT usage) in view of [this tentative DP note](https://indico.cern.ch/event/1722071/#6-part-ditau-path-in-phase-2-h)?

Yes, the plan is to provide continued support for DeepTau for phase-2 until its replacement is fully deployed. I think a safe estimate would be summer 2027. However, in my opinion, the main target of the migration from TF to ONNX are the Run3 legacy releases (17_0_X), for which DeepTau is required by definition. Does it make sense?

**Comment by mmusich** [comment]
> for which DeepTau is required by definition. Does it make sense?

as long as it gives bitwise same output it is acceptable for HLT to migrate to the new inference engine in the legacy releases (although I guess this goes beyond the scope of this issue, which AFAIU is about full removal of TF in the master release)

**Comment by makortel** [comment]
> (although I guess this goes beyond the scope of this issue, which AFAIU is about full removal of TF in the master release)

Correct. My take is that we are able to live with TF in 17_0_X (i.e. assume we won't have to update TF there), and I assume effort is spent better elsewhere than migrating away from TF for remaining Run 3 activities.

**Comment by mbluj** [comment]
> > (although I guess this goes beyond the scope of this issue, which AFAIU is about full removal of TF in the master release)
> 
> Correct. My take is that we are able to live with TF in 17_0_X (i.e. assume we won't have to update TF there), and I assume effort is spent better elsewhere than migrating away from TF for remaining Run 3 activities.

I generally agree, however it will not be a big effort to backport this migration to 17_0 with gain in performance and stability.

**Comment by pritampalit** [comment]
I made the PR : https://github.com/cms-sw/cmssw/pull/51885 . I added ONNX based CMSSW DeepTau producer and header and kept it totally separate from TF based producer and header, so that all TF based DeepTau can be safely dropped in future. I made the ONNX one as default (to be used in final processing) in python config.
